crates/ntx-builder/src/state.rs

```diff
@@ -0,0 +1,196 @@
+use std::{
+    collections::{BTreeMap, BTreeSet, VecDeque},
+    slice::Iter,
+};
+
+use miden_node_proto::domain::{mempool::MempoolEvent, note::NetworkNote};
+use miden_objects::{
+    Word,
+    account::{Account, AccountDelta, AccountId, delta::AccountUpdateDetails},
+    block::{BlockHeader, BlockNumber},
+    note::Nullifier,
+    transaction::{PartialBlockchain, TransactionId},
+};
+use miden_tx::{DataStore, DataStoreError, MastForestStore};
+
+use crate::builder::store::StoreClient;
+
+pub struct State {
+    latest_header: BlockHeader,
+    inflight: BTreeMap<TransactionId, InflightTx>,
+    accounts: BTreeMap<AccountId, AccountState>,
+    notes: BTreeMap<Nullifier, NetworkNote>,
+    nullifiers: BTreeSet<Nullifier>,
+}
+
+impl State {
+    pub fn new(genesis: BlockHeader) -> Self {
+        Self {
+            latest_header: genesis,
+            inflight: Default::default(),
+            accounts: Default::default(),
+            notes: Default::default(),
+            nullifiers: Default::default(),
+        }
+    }
+
+    /// Synchronizes the local committed state with that in the store.
+    pub async fn sync_committed(&mut self, _store: &StoreClient) -> anyhow::Result<()> {
+        todo!()
+    }
+
+    pub fn update(&mut self, event: MempoolEvent) {
+        match event {
+            MempoolEvent::TransactionAdded {
+                id,
+                nullifiers,
+                network_notes,
+                account_delta,
+            } => {
+                // TODO: AccountUpdateDetails is the wrong type to use.
+                //
+                // TODO: Update AccountState
+                let account_id = account_delta.map(|delta| match delta {
+                    AccountUpdateDetails::Private => unreachable!("Should never occur"),
+                    AccountUpdateDetails::New(account) => account.id(),
+                    AccountUpdateDetails::Delta(_account_delta) => {
+                        todo!("This is the wrong type to use here, we need an account ID")
```

Comment:
Is this outdated or do you still mean to change this to an `AccountId`? I guess not sure if the idea is to optimistically apply deltas after transactions are executed on the ntx builder or wait for them to be finalized in blocks (or doing something in the middle where you store various states for a single account and can then roll back appropriately and apply a new delta when finalized)

Reply:
`AccountDelta` doesn't currently contain account ID, but @PhilippGackstatter is planning to change this shortly.

Reply:
I'll wait for that change to land then.